### ISSUE TITLE
Push Based Consumer Acknowledgements

### ISSUE THREAD

**Issue by jon-whit** [nats-io/jetstream#312]
I'm experiencing an issue with a push based consumer using explicit acknowledgements. Here's a set of nats cli commands to reproduce the issue.

```
$ nats str add ORDERS -s nats://localhost:4222
? Subjects to consume ORDERS.*
? Storage backend file
? Retention Policy Interest
? Discard Policy Old
? Message count limit -1
? Message size limit -1
? Maximum message age limit -1
? Maximum individual message size -1
Stream ORDERS was created
Information for Stream ORDERS
Configuration:
             Subjects: ORDERS.*
     Acknowledgements: true
            Retention: File - Interest
             Replicas: 1
       Discard Policy: Old
     Maximum Messages: unlimited
        Maximum Bytes: unlimited
          Maximum Age: 0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited
State:
            Messages: 0
               Bytes: 0 B
            FirstSeq: 0
             LastSeq: 0
    Active Consumers: 0

// Publish to the ORDERS.scratch subject
$ nats pub ORDERS.scratch hello -s nats://localhost:4222 --wait
21:32:48 Published 5 bytes to ORDERS.scratch

// The stream is Interest based. Why does it have 1 message if there are 0 active consumers? This isn't inline with the Jetstream documentaton.
$ nats str info ORDERS -s nats://localhost:4222
Information for Stream ORDERS
Configuration:
             Subjects: ORDERS.*
     Acknowledgements: true
            Retention: File - Interest
             Replicas: 1
       Discard Policy: Old
     Maximum Messages: unlimited
        Maximum Bytes: unlimited
          Maximum Age: 0s
 Maximum Message Size: unlimited
    Maximum Consumers: unlimited
State:
            Messages: 1
               Bytes: 49 B
            FirstSeq: 1 @ 2020-08-27T21:32:48 UTC
             LastSeq: 1 @ 2020-08-27T21:32:48 UTC
    Active Consumers: 0

$ nats con ls ORDERS -s nats://localhost:4222
No Consumers defined

// Add a new consumer "NEW", deliver all subjects to `mytarget`, require explicit ack
$ nats con add -s nats://localhost:4222
? Consumer name NEW
? Delivery target mytarget
? Start policy (all, new, last, 1h, msg sequence) all
? Acknowledgement policy explicit
? Replay policy instant
? Filter Stream by subject (blank for all) 
? Maximum Allowed Deliveries 20
? Select a Stream ORDERS
Information for Consumer ORDERS > NEW
Configuration:
        Durable Name: NEW
    Delivery Subject: mytarget
         Deliver All: true
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
  Maximum Deliveries: 20
State:
  Last Delivered Message: Consumer sequence: 0 Stream sequence: 0
    Acknowledgment floor: Consumer sequence: 0 Stream sequence: 0
        Pending Messages: 0
    Redelivered Messages: 0

$ nats pub ORDERS.processed "order 1" --wait -s nats://localhost:4222
21:37:17 Sending request on [ORDERS.processed]
21:37:17 Received on [_INBOX.QU9nf0k5CoywZT7dqmmYHg]
+OK {"stream": "ORDERS", "seq": 2}

// Why are there 0 pending messages? Two messages have been published on the ORDERS.processed and ORDERS.scratch subjects. Shouldn't those be pending delivery to the target subscription subject 'mytarget'?
$ nats con info ORDERS NEW -s nats://localhost:4222
Information for Consumer ORDERS > NEW
Configuration:
        Durable Name: NEW
    Delivery Subject: mytarget
         Deliver All: true
          Ack Policy: Explicit
            Ack Wait: 30s
       Replay Policy: Instant
  Maximum Deliveries: 20
State:
  Last Delivered Message: Consumer sequence: 0 Stream sequence: 0
    Acknowledgment floor: Consumer sequence: 0 Stream sequence: 0
        Pending Messages: 0
    Redelivered Messages: 0

// Even though we `--ack` we still get redelivery of the message multiple times.. Why is this happening?
$ nats sub mytarget -s nats://localhost:4222 --raw --ack
hello
order 1
hello
order 1
hello
order 1
hello
order 1
```

What am I doing wrong? In general, what's the way to programmatically Acknowledge a message has been received by a consumer subscription? Using the `+ACK` reply message hasn't worked a single time for me, and I'm out of ideas on how to get this to work. It doesn't appear push based acknowledgements even work from my experiencing with them so far.

**Comment by ripienaar** [comment]
OK, so quite a lot going on here lets break it down.

First you note Interest based retention seems broken, and yes, to me it seems weird as well that when no consumers are configured that interest messages should be kept.  I opened https://github.com/nats-io/jetstream/issues/314 to track that seperately.

`Pending Messages` isnt the number of messages due to be delivered, it's the messages already attempted to be delivered but currently still waiting for some acknowledgement. 

Then regarding the acks, this is weird, if I just do `nats sub target` I too see the redliveries as expected, but adding `--ack` does acknowledge the messages and I do not get them again. Turns out this is a bug in `--raw` in the `nats sub` utility, I fixed that in https://github.com/nats-io/jetstream/pull/315
